File: QuoteGenerator.cs
// QuoteGenerator.cs - Генератор случайных цитат на C# (CLI)
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

namespace QuoteGenerator
{
    class Quote
    {
        public string Text { get; set; }
        public string Author { get; set; }
        public string Category { get; set; }
    }

    class Program
    {
        private static readonly string DATA_FILE = "quotes.json";
        private static readonly string FAV_FILE = "favorites.json";
        private static readonly Random random = new Random();

        private static List<Quote> DefaultQuotes() => new List<Quote>
        {
            new Quote { Text = "Жизнь — это то, что происходит с вами, пока вы строите планы.", Author = "Джон Леннон", Category = "мудрость" },
            new Quote { Text = "Будьте тем изменением, которое хотите видеть в мире.", Author = "Махатма Ганди", Category = "мотивация" },
            new Quote { Text = "Единственный способ делать отличную работу — любить то, что вы делаете.", Author = "Стив Джобс", Category = "успех" },
            new Quote { Text = "Счастье — это не цель, а способ путешествия.", Author = "Зигмунд Фрейд", Category = "жизнь" },
            new Quote { Text = "Тот, кто не рискует, тот не пьет шампанское.", Author = "Петр I", Category = "юмор" },
            new Quote { Text = "Лучший способ предсказать будущее — создать его.", Author = "Питер Друкер", Category = "успех" },
            new Quote { Text = "Тьма не может изгнать тьму: только свет может сделать это.", Author = "Мартин Лютер Кинг", Category = "мудрость" },
            new Quote { Text = "Не судите о книге по обложке.", Author = "Мигель де Сервантес", Category = "мудрость" },
            new Quote { Text = "Все, что мы есть, — это результат того, что мы думали.", Author = "Будда", Category = "мудрость" },
            new Quote { Text = "Самый трудный шаг — первый.", Author = "Китайская пословица", Category = "мотивация" },
            new Quote { Text = "Не бойтесь делать то, что не умеете. Помните, ковчег построил любитель, профессионалы построили «Титаник».", Author = "Дейв Барри", Category = "юмор" },
            new Quote { Text = "Всё, что вы можете вообразить, реально.", Author = "Пабло Пикассо", Category = "мотивация" },
            new Quote { Text = "Секрет успеха — в искренности. Как только вы сможете притвориться искренним, успех придёт.", Author = "Оскар Уайльд", Category = "юмор" },
            new Quote { Text = "Человек — это то, что он читает.", Author = "Джозеф Бродский", Category = "мудрость" },
            new Quote { Text = "Чтобы дойти до цели, надо прежде всего идти.", Author = "Оноре де Бальзак", Category = "мотивация" },
            new Quote { Text = "Учитесь правилам, чтобы нарушать их правильно.", Author = "Пабло Пикассо", Category = "мудрость" },
            new Quote { Text = "Судьба — это не вопрос случая, а вопрос выбора.", Author = "Уильям Дженнингс Брайан", Category = "жизнь" },
            new Quote { Text = "Никогда не сдавайтесь, потому что это просто место и время, когда течение повернётся в вашу пользу.", Author = "Гарриет Бичер-Стоу", Category = "мотивация" },
            new Quote { Text = "Счастье — это не отсутствие проблем, а умение с ними справляться.", Author = "Стив Мараболи", Category = "жизнь" },
            new Quote { Text = "Ваше время ограничено, не тратьте его на чужую жизнь.", Author = "Стив Джобс", Category = "успех" },
        };

        private static List<Quote> LoadQuotes()
        {
            if (File.Exists(DATA_FILE))
            {
                try
                {
                    string json = File.ReadAllText(DATA_FILE);
                    return JsonSerializer.Deserialize<List<Quote>>(json) ?? DefaultQuotes();
                }
                catch { }
            }
            var quotes = DefaultQuotes();
            SaveQuotes(quotes);
            return quotes;
        }

        private static void SaveQuotes(List<Quote> quotes)
        {
            string json = JsonSerializer.Serialize(quotes, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(DATA_FILE, json);
        }

        private static List<Quote> LoadFavorites()
        {
            if (File.Exists(FAV_FILE))
            {
                try
                {
                    string json = File.ReadAllText(FAV_FILE);
                    return JsonSerializer.Deserialize<List<Quote>>(json) ?? new List<Quote>();
                }
                catch { }
            }
            return new List<Quote>();
        }

        private static void SaveFavorites(List<Quote> favs)
        {
            string json = JsonSerializer.Serialize(favs, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(FAV_FILE, json);
        }

        private static Quote GetRandomQuote(List<Quote> quotes)
        {
            if (quotes.Count == 0) return null;
            return quotes[random.Next(quotes.Count)];
        }

        private static List<Quote> SearchQuotes(List<Quote> quotes, string keyword)
        {
            keyword = keyword.ToLower();
            return quotes.Where(q => q.Text.ToLower().Contains(keyword) || q.Author.ToLower().Contains(keyword)).ToList();
        }

        private static Quote AddQuote(List<Quote> quotes, string text, string author, string category)
        {
            if (string.IsNullOrEmpty(category)) category = "общее";
            var q = new Quote { Text = text, Author = author, Category = category };
            quotes.Add(q);
            SaveQuotes(quotes);
            return q;
        }

        private static string ExportQuote(Quote q, string filename)
        {
            if (string.IsNullOrEmpty(filename))
                filename = $"quote_{DateTime.Now:yyyyMMdd_HHmmss}.txt";
            File.WriteAllText(filename, $"\"{q.Text}\"\n— {q.Author}\nКатегория: {q.Category}");
            return filename;
        }

        private static List<string> GetCategories(List<Quote> quotes)
        {
            return quotes.Select(q => q.Category).Distinct().OrderBy(c => c).ToList();
        }

        private static List<Quote> FilterByCategory(List<Quote> quotes, string category)
        {
            return quotes.Where(q => q.Category == category).ToList();
        }

        private static bool Contains(List<string> list, string item)
        {
            return list.Contains(item);
        }

        private static void Interactive()
        {
            Console.WriteLine("💬 ГЕНЕРАТОР СЛУЧАЙНЫХ ЦИТАТ");
            var quotes = LoadQuotes();
            var favorites = LoadFavorites();
            while (true)
            {
                Console.WriteLine("\nВыберите действие:");
                Console.WriteLine("1. Показать случайную цитату");
                Console.WriteLine("2. Поиск цитат");
                Console.WriteLine("3. Добавить цитату");
                Console.WriteLine("4. Избранное");
                Console.WriteLine("5. Экспорт цитаты");
                Console.WriteLine("6. Категории");
                Console.WriteLine("7. Случайная из категории");
                Console.WriteLine("0. Выход");
                Console.Write("Ваш выбор: ");
                string choice = Console.ReadLine();
                if (choice == "0") break;
                else if (choice == "1")
                {
                    var q = GetRandomQuote(quotes);
                    if (q != null)
                    {
                        Console.WriteLine($"\n\"{q.Text}\"\n— {q.Author} (Категория: {q.Category})");
                        Console.Write("Добавить в избранное? (y/n): ");
                        if (Console.ReadLine()?.ToLower() == "y")
                        {
                            if (!favorites.Any(f => f.Text == q.Text && f.Author == q.Author))
                            {
                                favorites.Add(q);
                                SaveFavorites(favorites);
                                Console.WriteLine("✅ Добавлено в избранное!");
                            }
                            else Console.WriteLine("Уже в избранном.");
                        }
                    }
                    else Console.WriteLine("Нет цитат.");
                }
                else if (choice == "2")
                {
                    Console.Write("Введите ключевое слово или автора: ");
                    string keyword = Console.ReadLine();
                    if (string.IsNullOrEmpty(keyword)) { Console.WriteLine("Введите запрос."); continue; }
                    var results = SearchQuotes(quotes, keyword);
                    if (results.Count == 0)
                    {
                        Console.WriteLine("Ничего не найдено.");
                    }
                    else
                    {
                        Console.WriteLine($"\nНайдено {results.Count} цитат:");
                        for (int i = 0; i < results.Count; i++)
                        {
                            var q = results[i];
                            Console.WriteLine($"{i+1}. \"{q.Text}\" — {q.Author}");
                        }
                        Console.Write("Введите номер, чтобы сохранить в избранное (или Enter для пропуска): ");
                        string idxStr = Console.ReadLine();
                        if (!string.IsNullOrEmpty(idxStr) && int.TryParse(idxStr, out int idx) && idx >= 1 && idx <= results.Count)
                        {
                            var q = results[idx-1];
                            if (!favorites.Any(f => f.Text == q.Text && f.Author == q.Author))
                            {
                                favorites.Add(q);
                                SaveFavorites(favorites);
                                Console.WriteLine("✅ Добавлено в избранное!");
                            }
                            else Console.WriteLine("Уже в избранном.");
                        }
                    }
                }
                else if (choice == "3")
                {
                    Console.Write("Введите текст цитаты: ");
                    string text = Console.ReadLine();
                    if (string.IsNullOrEmpty(text)) { Console.WriteLine("Текст обязателен."); continue; }
                    Console.Write("Введите автора: ");
                    string author = Console.ReadLine();
                    if (string.IsNullOrEmpty(author)) author = "Неизвестен";
                    Console.Write("Введите категорию (по умолчанию общее): ");
                    string cat = Console.ReadLine();
                    if (string.IsNullOrEmpty(cat)) cat = "общее";
                    var q = AddQuote(quotes, text, author, cat);
                    Console.WriteLine($"✅ Цитата добавлена: \"{q.Text}\" — {q.Author}");
                }
                else if (choice == "4")
                {
                    if (favorites.Count == 0)
                    {
                        Console.WriteLine("Избранное пусто.");
                    }
                    else
                    {
                        Console.WriteLine("\n⭐ ИЗБРАННОЕ:");
                        for (int i = 0; i < favorites.Count; i++)
                        {
                            var q = favorites[i];
                            Console.WriteLine($"{i+1}. \"{q.Text}\" — {q.Author}");
                        }
                        Console.Write("Введите номер, чтобы удалить из избранного (или Enter для пропуска): ");
                        string idxStr = Console.ReadLine();
                        if (!string.IsNullOrEmpty(idxStr) && int.TryParse(idxStr, out int idx) && idx >= 1 && idx <= favorites.Count)
                        {
                            var removed = favorites[idx-1];
                            favorites.RemoveAt(idx-1);
                            SaveFavorites(favorites);
                            Console.WriteLine($"Удалено: \"{removed.Text}\"");
                        }
                    }
                }
                else if (choice == "5")
                {
                    var q = GetRandomQuote(quotes);
                    if (q == null) { Console.WriteLine("Нет цитат для экспорта."); continue; }
                    Console.Write("Имя файла (по умолчанию создастся автоматически): ");
                    string filename = Console.ReadLine();
                    string fname = ExportQuote(q, string.IsNullOrEmpty(filename) ? null : filename);
                    Console.WriteLine($"Цитата экспортирована в {fname}");
                }
                else if (choice == "6")
                {
                    var cats = GetCategories(quotes);
                    if (cats.Count == 0)
                    {
                        Console.WriteLine("Нет категорий.");
                    }
                    else
                    {
                        Console.WriteLine("\n📂 ДОСТУПНЫЕ КАТЕГОРИИ:");
                        foreach (var cat in cats)
                        {
                            int count = quotes.Count(q => q.Category == cat);
                            Console.WriteLine($"  {cat} ({count} цитат)");
                        }
                        Console.Write("Введите категорию для просмотра цитат (или Enter для пропуска): ");
                        string catChoice = Console.ReadLine();
                        if (!string.IsNullOrEmpty(catChoice) && Contains(cats, catChoice))
                        {
                            var filtered = FilterByCategory(quotes, catChoice);
                            Console.WriteLine($"\nЦитаты в категории '{catChoice}':");
                            for (int i = 0; i < Math.Min(10, filtered.Count); i++)
                            {
                                var q = filtered[i];
                                Console.WriteLine($"{i+1}. \"{q.Text}\" — {q.Author}");
                            }
                            if (filtered.Count > 10)
                                Console.WriteLine($"... и ещё {filtered.Count-10} цитат.");
                        }
                    }
                }
                else if (choice == "7")
                {
                    var cats = GetCategories(quotes);
                    if (cats.Count == 0)
                    {
                        Console.WriteLine("Нет категорий.");
                        continue;
                    }
                    Console.WriteLine("Доступные категории: " + string.Join(", ", cats));
                    Console.Write("Введите категорию: ");
                    string cat = Console.ReadLine();
                    if (string.IsNullOrEmpty(cat) || !Contains(cats, cat))
                    {
                        Console.WriteLine("Категория не найдена.");
                        continue;
                    }
                    var filtered = FilterByCategory(quotes, cat);
                    if (filtered.Count == 0)
                    {
                        Console.WriteLine("В этой категории нет цитат.");
                    }
                    else
                    {
                        var q = filtered[random.Next(filtered.Count)];
                        Console.WriteLine($"\n\"{q.Text}\"\n— {q.Author} (Категория: {cat})");
                    }
                }
                else
                {
                    Console.WriteLine("Неверный выбор.");
                }
            }
        }

        static void Main(string[] args)
        {
            if (args.Length > 0)
            {
                string mode = args[0];
                var quotes = LoadQuotes();
                if (mode == "--random")
                {
                    var q = GetRandomQuote(quotes);
                    if (q != null) Console.WriteLine($"\"{q.Text}\" — {q.Author}");
                }
                else if (mode == "--search" && args.Length > 1)
                {
                    var results = SearchQuotes(quotes, args[1]);
                    foreach (var q in results) Console.WriteLine($"\"{q.Text}\" — {q.Author}");
                }
                else if (mode == "--add" && args.Length > 2)
                {
                    string author = args.Length > 3 ? args[3] : "Неизвестен";
                    string cat = args.Length > 4 ? args[4] : "общее";
                    var q = AddQuote(quotes, args[1], author, cat);
                    Console.WriteLine($"Добавлено: \"{q.Text}\" — {q.Author}");
                }
                else if (mode == "--export")
                {
                    var q = GetRandomQuote(quotes);
                    if (q != null)
                    {
                        string fname = ExportQuote(q, args.Length > 1 ? args[1] : null);
                        Console.WriteLine($"Экспортировано в {fname}");
                    }
                }
                else if (mode == "--favorites")
                {
                    var favs = LoadFavorites();
                    foreach (var q in favs) Console.WriteLine($"\"{q.Text}\" — {q.Author}");
                }
                else
                {
                    Console.WriteLine("Использование: QuoteGenerator [--random] [--search KEYWORD] [--add TEXT AUTHOR [CATEGORY]] [--export [FILE]] [--favorites]");
                }
            }
            else
            {
                Interactive();
            }
        }
    }
}

File: quote_generator.rs
// quote_generator.rs - Генератор случайных цитат на Rust (CLI)
use serde::{Serialize, Deserialize};
use std::collections::HashSet;
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::Path;
use std::str::FromStr;
use rand::seq::SliceRandom;
use rand::thread_rng;
use chrono::Local;

#[derive(Serialize, Deserialize, Clone)]
struct Quote {
    text: String,
    author: String,
    category: String,
}

const DATA_FILE: &str = "quotes.json";
const FAV_FILE: &str = "favorites.json";

fn default_quotes() -> Vec<Quote> {
    vec![
        Quote { text: "Жизнь — это то, что происходит с вами, пока вы строите планы.".to_string(), author: "Джон Леннон".to_string(), category: "мудрость".to_string() },
        Quote { text: "Будьте тем изменением, которое хотите видеть в мире.".to_string(), author: "Махатма Ганди".to_string(), category: "мотивация".to_string() },
        Quote { text: "Единственный способ делать отличную работу — любить то, что вы делаете.".to_string(), author: "Стив Джобс".to_string(), category: "успех".to_string() },
        Quote { text: "Счастье — это не цель, а способ путешествия.".to_string(), author: "Зигмунд Фрейд".to_string(), category: "жизнь".to_string() },
        Quote { text: "Тот, кто не рискует, тот не пьет шампанское.".to_string(), author: "Петр I".to_string(), category: "юмор".to_string() },
        Quote { text: "Лучший способ предсказать будущее — создать его.".to_string(), author: "Питер Друкер".to_string(), category: "успех".to_string() },
        Quote { text: "Тьма не может изгнать тьму: только свет может сделать это.".to_string(), author: "Мартин Лютер Кинг".to_string(), category: "мудрость".to_string() },
        Quote { text: "Не судите о книге по обложке.".to_string(), author: "Мигель де Сервантес".to_string(), category: "мудрость".to_string() },
        Quote { text: "Все, что мы есть, — это результат того, что мы думали.".to_string(), author: "Будда".to_string(), category: "мудрость".to_string() },
        Quote { text: "Самый трудный шаг — первый.".to_string(), author: "Китайская пословица".to_string(), category: "мотивация".to_string() },
        Quote { text: "Не бойтесь делать то, что не умеете. Помните, ковчег построил любитель, профессионалы построили «Титаник».".to_string(), author: "Дейв Барри".to_string(), category: "юмор".to_string() },
        Quote { text: "Всё, что вы можете вообразить, реально.".to_string(), author: "Пабло Пикассо".to_string(), category: "мотивация".to_string() },
        Quote { text: "Секрет успеха — в искренности. Как только вы сможете притвориться искренним, успех придёт.".to_string(), author: "Оскар Уайльд".to_string(), category: "юмор".to_string() },
        Quote { text: "Человек — это то, что он читает.".to_string(), author: "Джозеф Бродский".to_string(), category: "мудрость".to_string() },
        Quote { text: "Чтобы дойти до цели, надо прежде всего идти.".to_string(), author: "Оноре де Бальзак".to_string(), category: "мотивация".to_string() },
        Quote { text: "Учитесь правилам, чтобы нарушать их правильно.".to_string(), author: "Пабло Пикассо".to_string(), category: "мудрость".to_string() },
        Quote { text: "Судьба — это не вопрос случая, а вопрос выбора.".to_string(), author: "Уильям Дженнингс Брайан".to_string(), category: "жизнь".to_string() },
        Quote { text: "Никогда не сдавайтесь, потому что это просто место и время, когда течение повернётся в вашу пользу.".to_string(), author: "Гарриет Бичер-Стоу".to_string(), category: "мотивация".to_string() },
        Quote { text: "Счастье — это не отсутствие проблем, а умение с ними справляться.".to_string(), author: "Стив Мараболи".to_string(), category: "жизнь".to_string() },
        Quote { text: "Ваше время ограничено, не тратьте его на чужую жизнь.".to_string(), author: "Стив Джобс".to_string(), category: "успех".to_string() },
    ]
}

fn load_quotes() -> Vec<Quote> {
    if Path::new(DATA_FILE).exists() {
        if let Ok(data) = fs::read_to_string(DATA_FILE) {
            if let Ok(quotes) = serde_json::from_str(&data) {
                return quotes;
            }
        }
    }
    let quotes = default_quotes();
    save_quotes(&quotes);
    quotes
}

fn save_quotes(quotes: &[Quote]) {
    let data = serde_json::to_string_pretty(quotes).unwrap();
    fs::write(DATA_FILE, data).unwrap();
}

fn load_favorites() -> Vec<Quote> {
    if Path::new(FAV_FILE).exists() {
        if let Ok(data) = fs::read_to_string(FAV_FILE) {
            if let Ok(favs) = serde_json::from_str(&data) {
                return favs;
            }
        }
    }
    Vec::new()
}

fn save_favorites(favs: &[Quote]) {
    let data = serde_json::to_string_pretty(favs).unwrap();
    fs::write(FAV_FILE, data).unwrap();
}

fn get_random_quote(quotes: &[Quote]) -> Option<Quote> {
    if quotes.is_empty() {
        None
    } else {
        let mut rng = thread_rng();
        quotes.choose(&mut rng).cloned()
    }
}

fn search_quotes(quotes: &[Quote], keyword: &str) -> Vec<Quote> {
    let keyword = keyword.to_lowercase();
    quotes.iter()
        .filter(|q| q.text.to_lowercase().contains(&keyword) || q.author.to_lowercase().contains(&keyword))
        .cloned()
        .collect()
}

fn add_quote(quotes: &mut Vec<Quote>, text: &str, author: &str, category: &str) -> Quote {
    let q = Quote {
        text: text.to_string(),
        author: if author.is_empty() { "Неизвестен".to_string() } else { author.to_string() },
        category: if category.is_empty() { "общее".to_string() } else { category.to_string() },
    };
    quotes.push(q.clone());
    save_quotes(quotes);
    q
}

fn export_quote(q: &Quote, filename: Option<&str>) -> String {
    let filename = match filename {
        Some(f) => f.to_string(),
        None => format!("quote_{}.txt", Local::now().format("%Y%m%d_%H%M%S")),
    };
    let content = format!("\"{}\"\n— {}\nКатегория: {}", q.text, q.author, q.category);
    fs::write(&filename, content).unwrap();
    filename
}

fn get_categories(quotes: &[Quote]) -> Vec<String> {
    let mut set = HashSet::new();
    for q in quotes {
        set.insert(q.category.clone());
    }
    let mut cats: Vec<_> = set.into_iter().collect();
    cats.sort();
    cats
}

fn filter_by_category(quotes: &[Quote], category: &str) -> Vec<Quote> {
    quotes.iter().filter(|q| q.category == category).cloned().collect()
}

fn contains(slice: &[String], item: &str) -> bool {
    slice.iter().any(|s| s == item)
}

fn read_line(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn interactive() {
    println!("💬 ГЕНЕРАТОР СЛУЧАЙНЫХ ЦИТАТ");
    let mut quotes = load_quotes();
    let mut favorites = load_favorites();
    let stdin = io::stdin();
    loop {
        println!("\nВыберите действие:");
        println!("1. Показать случайную цитату");
        println!("2. Поиск цитат");
        println!("3. Добавить цитату");
        println!("4. Избранное");
        println!("5. Экспорт цитаты");
        println!("6. Категории");
        println!("7. Случайная из категории");
        println!("0. Выход");
        let choice = read_line("Ваш выбор: ");
        match choice.as_str() {
            "0" => break,
            "1" => {
                if let Some(q) = get_random_quote(&quotes) {
                    println!("\n\"{}\"\n— {} (Категория: {})", q.text, q.author, q.category);
                    let ans = read_line("Добавить в избранное? (y/n): ");
                    if ans.to_lowercase() == "y" {
                        if !favorites.iter().any(|f| f.text == q.text && f.author == q.author) {
                            favorites.push(q);
                            save_favorites(&favorites);
                            println!("✅ Добавлено в избранное!");
                        } else {
                            println!("Уже в избранном.");
                        }
                    }
                } else {
                    println!("Нет цитат.");
                }
            }
            "2" => {
                let keyword = read_line("Введите ключевое слово или автора: ");
                if keyword.is_empty() {
                    println!("Введите запрос.");
                    continue;
                }
                let results = search_quotes(&quotes, &keyword);
                if results.is_empty() {
                    println!("Ничего не найдено.");
                } else {
                    println!("\nНайдено {} цитат:", results.len());
                    for (i, q) in results.iter().enumerate() {
                        println!("{}. \"{}\" — {}", i+1, q.text, q.author);
                    }
                    let idx_str = read_line("Введите номер, чтобы сохранить в избранное (или Enter для пропуска): ");
                    if let Ok(idx) = idx_str.parse::<usize>() {
                        if idx >= 1 && idx <= results.len() {
                            let q = results[idx-1].clone();
                            if !favorites.iter().any(|f| f.text == q.text && f.author == q.author) {
                                favorites.push(q);
                                save_favorites(&favorites);
                                println!("✅ Добавлено в избранное!");
                            } else {
                                println!("Уже в избранном.");
                            }
                        }
                    }
                }
            }
            "3" => {
                let text = read_line("Введите текст цитаты: ");
                if text.is_empty() {
                    println!("Текст обязателен.");
                    continue;
                }
                let author = read_line("Введите автора: ");
                let author = if author.is_empty() { "Неизвестен".to_string() } else { author };
                let category = read_line("Введите категорию (по умолчанию общее): ");
                let category = if category.is_empty() { "общее".to_string() } else { category };
                let q = add_quote(&mut quotes, &text, &author, &category);
                println!("✅ Цитата добавлена: \"{}\" — {}", q.text, q.author);
            }
            "4" => {
                if favorites.is_empty() {
                    println!("Избранное пусто.");
                } else {
                    println!("\n⭐ ИЗБРАННОЕ:");
                    for (i, q) in favorites.iter().enumerate() {
                        println!("{}. \"{}\" — {}", i+1, q.text, q.author);
                    }
                    let idx_str = read_line("Введите номер, чтобы удалить из избранного (или Enter для пропуска): ");
                    if let Ok(idx) = idx_str.parse::<usize>() {
                        if idx >= 1 && idx <= favorites.len() {
                            let removed = favorites.remove(idx-1);
                            save_favorites(&favorites);
                            println!("Удалено: \"{}\"", removed.text);
                        }
                    }
                }
            }
            "5" => {
                if let Some(q) = get_random_quote(&quotes) {
                    let filename = read_line("Имя файла (по умолчанию создастся автоматически): ");
                    let fname = if filename.is_empty() {
                        export_quote(&q, None)
                    } else {
                        export_quote(&q, Some(&filename))
                    };
                    println!("Цитата экспортирована в {}", fname);
                } else {
                    println!("Нет цитат для экспорта.");
                }
            }
            "6" => {
                let cats = get_categories(&quotes);
                if cats.is_empty() {
                    println!("Нет категорий.");
                } else {
                    println!("\n📂 ДОСТУПНЫЕ КАТЕГОРИИ:");
                    for cat in &cats {
                        let count = quotes.iter().filter(|q| q.category == *cat).count();
                        println!("  {} ({} цитат)", cat, count);
                    }
                    let cat_choice = read_line("Введите категорию для просмотра цитат (или Enter для пропуска): ");
                    if !cat_choice.is_empty() && contains(&cats, &cat_choice) {
                        let filtered = filter_by_category(&quotes, &cat_choice);
                        println!("\nЦитаты в категории '{}':", cat_choice);
                        for (i, q) in filtered.iter().enumerate().take(10) {
                            println!("{}. \"{}\" — {}", i+1, q.text, q.author);
                        }
                        if filtered.len() > 10 {
                            println!("... и ещё {} цитат.", filtered.len() - 10);
                        }
                    }
                }
            }
            "7" => {
                let cats = get_categories(&quotes);
                if cats.is_empty() {
                    println!("Нет категорий.");
                    continue;
                }
                println!("Доступные категории: {}", cats.join(", "));
                let cat = read_line("Введите категорию: ");
                if cat.is_empty() || !contains(&cats, &cat) {
                    println!("Категория не найдена.");
                    continue;
                }
                let filtered = filter_by_category(&quotes, &cat);
                if filtered.is_empty() {
                    println!("В этой категории нет цитат.");
                } else {
                    if let Some(q) = filtered.choose(&mut thread_rng()) {
                        println!("\n\"{}\"\n— {} (Категория: {})", q.text, q.author, q.category);
                    }
                }
            }
            _ => println!("Неверный выбор."),
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 {
        let mut random = false;
        let mut search = String::new();
        let mut add_text = String::new();
        let mut add_author = String::new();
        let mut category = String::new();
        let mut export = false;
        let mut favorites = false;
        let mut i = 1;
        while i < args.len() {
            match args[i].as_str() {
                "--random" => random = true,
                "--search" => { search = args[i+1].clone(); i += 1; }
                "--add" => { add_text = args[i+1].clone(); i += 1; }
                "--author" => { add_author = args[i+1].clone(); i += 1; }
                "--category" => { category = args[i+1].clone(); i += 1; }
                "--export" => export = true,
                "--favorites" => favorites = true,
                _ => {}
            }
            i += 1;
        }
        let quotes = load_quotes();
        if random {
            if let Some(q) = get_random_quote(&quotes) {
                println!("\"{}\" — {}", q.text, q.author);
            }
        } else if !search.is_empty() {
            let results = search_quotes(&quotes, &search);
            for q in results {
                println!("\"{}\" — {}", q.text, q.author);
            }
        } else if !add_text.is_empty() && !add_author.is_empty() {
            let mut qs = quotes;
            let q = add_quote(&mut qs, &add_text, &add_author, if category.is_empty() { "общее" } else { &category });
            println!("Добавлено: \"{}\" — {}", q.text, q.author);
        } else if export {
            if let Some(q) = get_random_quote(&quotes) {
                let fname = export_quote(&q, None);
                println!("Экспортировано в {}", fname);
            }
        } else if favorites {
            let favs = load_favorites();
            for q in favs {
                println!("\"{}\" — {}", q.text, q.author);
            }
        } else {
            interactive();
        }
    } else {
        interactive();
    }
}
